### PR TITLE
feat: whatsapp duplicate message

### PR DESCRIPTION
**Pull request by tejaswinichile**
Fixes: https://linear.app/chatwoot/issue/CW-1525/duplicate-messages-in-whatsapp-conversations

### CODE REVIEW

File: app/services/whatsapp/incoming_message_base_service.rb [hunk 1]
```diff
@@ -47,17 +48,14 @@ def update_message_with_status(message, status)
   end
 
   def create_messages
+    ::Redis::Alfred.delete(@processed_params[:messages].first[:id])
```
Comment (pranavrajs) [inline app/services/whatsapp/incoming_message_base_service.rb:51]: I would suggest deleting this after the message is created.

File: app/services/whatsapp/incoming_message_base_service.rb [hunk 2]
```diff
@@ -20,13 +20,15 @@ def perform
 
   def process_messages
     # message allready exists so we don't need to process
-    return if find_message_by_source_id(@processed_params[:messages].first[:id])
+    return if find_message_by_source_id(@processed_params[:messages].first[:id]) || message_under_process?
 
+    set_message_source_id
```
Comment (sojan-official) [inline app/services/whatsapp/incoming_message_base_service.rb:25]:
```suggestion
    cache_message_source_id_in_redis
```

File: app/services/whatsapp/incoming_message_base_service.rb [hunk 3]
```diff
@@ -20,13 +20,15 @@ def perform
...
     set_contact
     return unless @contact
 
     set_conversation
     create_messages
+    delete_message_source_id
```
Comment (sojan-official) [inline app/services/whatsapp/incoming_message_base_service.rb:31]:
```suggestion
    clear_message_source_id_from_redis
```

File: app/services/whatsapp/incoming_message_service_helpers.rb [hunk 4]
```diff
@@ -104,4 +104,16 @@ def find_message_by_source_id(source_id)
 
     @message = Message.find_by(source_id: source_id)
   end
+
+  def message_under_process?
+    Redis::Alfred.get(@processed_params[:messages].first[:id])
+  end
+
+  def set_message_source_id
+    ::Redis::Alfred.setex(@processed_params[:messages].first[:id], true) if @processed_params.try(:[], :messages).present?
```
Comment (sojan-official) [inline app/services/whatsapp/incoming_message_service_helpers.rb:113]: Lets use a namespace Redis key here : 
Example: https://github.com/chatwoot/chatwoot/blob/847d7ea082ac6761d3949fd6209f885c5925acf5/lib/redis/redis_keys.rb#L10

will help us to group / target specific keys related to this fix alone in future.

**Review by sojan-official — COMMENTED** [review]
The changes look good to me. Some cosmetic suggestions for better readability

File: app/services/whatsapp/incoming_message_service_helpers.rb (hunk 4 above)
Comment (tejaswinichile) [reply app/services/whatsapp/incoming_message_service_helpers.rb]: @sojan-official Done with the changes. Please have a look.

**Review by sojan-official — APPROVED** [review]

**Pull request merged**